edn/src/pp.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::borrow::Cow;
+use std::io;
+use pretty;
+
+/// A sum type representing printing instructions in a pretty print tree.
+#[derive(Debug)]
+pub enum PrintInstr<'a> {
+    /// Print a newline.
+    PrintNewline,
+    /// Print a string.
+    PrintString(Cow<'a, str>),
+    /// Print a newline, the children in the tree as a continuous block
+    /// nested at `usize`, then another newline.
+    NestChildren(usize),
+}
+
+/// A tree of pretty print instructions.
```

Comment:
Can you explain _why_ you need this intermediate representation?  That is, you're converting `edn::Value` -> `PrettyPrintTree` -> `pretty::DocBuilder` -> `String`.  What value is `PrettyPrintTree` capturing?

I expected this patch to be the real pretty printing bits (that handle `edn::Value`) that accepts an allocator and just uses it, like https://github.com/freebroccolo/pretty.rs/blob/master/examples/trees.rs#L23.  Can you explain why that approach isn't sufficient?